src/main/java/fr/mrmicky/fastparticle/compatibility/ParticleSenderLegacy.java

```diff
@@ -0,0 +1,228 @@
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.logging.Level;
+
+/**
+ * Legacy particle sender with NMS for 1.7/1.8 servers
+ *
+ * @author MrMicky
+ */
+@SuppressWarnings("deprecation")
+public class ParticleSenderLegacy extends AbstractParticleSender {
+
+    private static final boolean SERVER_IS_1_8;
+    private static final String PACKAGE_NAME_NMS;
+    private static final String PACKAGE_NAME_OCB;
+
+    private static final Constructor<?> PACKET_PARTICLE;
+    private static final Class<?> ENUM_PARTICLE;
+
+    private static final Method WORLD_GET_HANDLE;
+    private static final Method WORLD_SEND_PARTICLE;
+
+    private static final Method PLAYER_GET_HANDLE;
+    private static final Field PLAYER_CONNECTION;
+    private static final Method SEND_PACKET;
+
+    static {
+        String ver = FastParticle.SERVER_VERSION;
+
+        PACKAGE_NAME_NMS = "net.minecraft.server." + ver;
+        PACKAGE_NAME_OCB = "org.bukkit.craftbukkit." + ver;
+        SERVER_IS_1_8 = ver.startsWith("v1_8_");
+
+        Constructor<?> packetParticle = null;
+        Class<?> enumParticle = null;
+
+        Method worldGetHandle = null;
+        Method worldSendParticle = null;
+
+        Method playerGetHandle = null;
+        Field playerConnection = null;
+        Method sendPacket = null;
+
+        try {
+            Class<?> packetParticleClass = getClassNMS("PacketPlayOutWorldParticles");
+            Class<?> playerClass = getClassNMS("EntityPlayer");
+            Class<?> playerConnectionClass = getClassNMS("PlayerConnection");
+            Class<?> packetClass = getClassNMS("Packet");
+            Class<?> worldClass = getClassNMS("WorldServer");
+            Class<?> entityPlayerClass = getClassNMS("EntityPlayer");
+
+            Class<?> craftPlayerClass = getClassOCB("entity.CraftPlayer");
+            Class<?> craftWorldClass = getClassOCB("CraftWorld");
+
+            if (SERVER_IS_1_8) {
+                enumParticle = getClassNMS("EnumParticle");
+                packetParticle = packetParticleClass.getConstructor(enumParticle, boolean.class, float.class,
+                        float.class, float.class, float.class, float.class, float.class, float.class, int.class,
+                        int[].class);
+                worldSendParticle = worldClass.getDeclaredMethod("sendParticles", entityPlayerClass, enumParticle,
+                        boolean.class, double.class, double.class, double.class, int.class, double.class, double.class,
+                        double.class, double.class, int[].class);
+            } else {
+                packetParticle = packetParticleClass.getConstructor(String.class, float.class, float.class, float.class,
+                        float.class, float.class, float.class, float.class, int.class);
+                worldSendParticle = worldClass.getDeclaredMethod("a", String.class, double.class, double.class,

```

Comment:
isn't this going to break with each new server version? I mean, these obfuscated methods tend to randomly change on every slight change in the server, or am I wrong? is there any way around this?

Reply:
The ParticleSenderLegacy is only for 1.7 and 1.8 servers, and this method is only for 1.7 servers so it will not break with updates